lib/log_dna_batcher.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+require "net/http"
+require "uri"
+require "json"
+require "socket"
+
+class LogDnaBatcher
+  def initialize(api_key, base_url: "https://logs.logdna.com/logs/ingest",
+    default_metadata: {"app" => $0, "pid" => Process.pid},
+    flush_interval: 0.5,
+    max_batch_size: 5)
+    @api_key = api_key
+    @flush_interval = flush_interval
+    @max_batch_size = max_batch_size
+    @input_queue = Queue.new
+    @http = nil
+    @base_url = URI.parse(base_url)
+    @default_metadata = default_metadata.dup.freeze
+
+    params = @base_url.query ? URI.decode_www_form(@base_url.query).to_h : {}
+    params["hostname"] = Socket.gethostname
+    params["app"] = $0
+    @request_path = "#{@base_url.path}?#{URI.encode_www_form(params)}"
+
+    start_processor
+  end
+
+  def log(line, timestamp: (Time.now.to_f * 1000).to_i, **metadata)
+    data = @default_metadata.dup
+    data.merge!(metadata)
+    data[:line] = line
+    data[:timestamp] = timestamp
+    @input_queue.push(data)
+    nil
+  end
+
+  def stop
+    @input_queue.close
+    @processor_thread&.join
+    close_connection
+  end
+
+  def send_batch(batch)
+    return if batch.empty?
+
+    begin
+      http = ensure_connection
+
+      request = Net::HTTP::Post.new(@request_path)
+      request.basic_auth(@api_key, "")
+      request["Content-Type"] = "application/json; charset=UTF-8"
+      request["Connection"] = "keep-alive"
+      request.body = {lines: batch}.to_json
+
+      response = http.request(request)
+
+      if response.is_a?(Net::HTTPSuccess)
+        batch.clear
+      else
+        puts "Failed to send logs: #{response.code} #{response.message}"
+      end
+    rescue => e
+      puts "Error sending batch: #{e.message}"
+      close_connection
+    end
+  end
+
+  def ensure_connection
+    return @http if @http&.started?
+
+    close_connection if @http
+
+    base_uri = URI.parse(@base_url)
+    @http = Net::HTTP.new(base_uri.host, base_uri.port)
+    @http.use_ssl = base_uri.scheme == "https"
+    @http.keep_alive_timeout = 30
+    @http.start
+    @http
+  end
+
+  private
+
+  def start_processor
+    @processor_thread = Thread.new do
+      batch = []
+      last_flush_time = Time.now
+
+      loop do
+        if (item = @input_queue.pop(timeout: @flush_interval))
+          batch << item
+        end
+
+        input_queue_is_closed = @input_queue.closed? && @input_queue.empty?
+        flush_interval_is_exceeded = (Time.now - last_flush_time) >= @flush_interval
+        max_batch_size_is_exceeded = batch.size >= @max_batch_size
+
+        if input_queue_is_closed || flush_interval_is_exceeded || max_batch_size_is_exceeded
+          send_batch(batch)
+        end
+
+        if batch.empty?
+          last_flush_time = Time.now
+          break if input_queue_is_closed
+        end
+      rescue => e
+        puts "Error in processor: #{e.message}"
```

Comment:
The error message here should be more specific, since it could show up on the user's terminal. I would mention the class and method.

In production, if we hit this, we may have already missed logging something we should have logged. I don't think we should continue processing at that point, because we could continue missing logs. We should probably have the `rescue` at the thread level, and should exit the process after printing the error message.